astropy/utils/data.py

```diff
@@ -1545,14 +1554,45 @@ def _keep_trying(timeout):
 # access to the shelve even for reading.
 @contextlib.contextmanager
 def _cache_lock(pkgname, need_write=False):
-    lockdir = os.path.join(_get_download_cache_locs(pkgname)[0], 'lock')
-    pidfn = os.path.join(lockdir, 'pid')
-    got_lock = False
-    try:
+    cache_dir = _get_download_cache_locs(pkgname)[0]
+    lock_name = 'lock'
+    lock_dir = os.path.join(cache_dir, lock_name)
+
+    def rmdir_fd(fd):
+        # Remove the lock directory given a file descriptor
+        # pointing to the cache directory.
+        try:
+            lock_fd = os.open(lock_name, os.O_RDONLY, dir_fd=fd)
```

Comment:
Interesting. It is almost counter-intuitive (at least for me) to first open the descriptor to the directory before removing it. Would you mind add some comments to explain why this is done?

Reply:
Good point. The dir_fd interface for race-free filesystem manipulation is unfamiliar, not portable, and awkward. But it might fix some nasty concurrency problems. I'll elaborate.

Reply:
Er, actually, it's wrong, or rather, a vestige of when I was still making and deleting PID files. If I add that back in it should be in a context manager anyway. Sorry.